sys/src/cmd/gdbserver/gdbstub.c

```diff
@@ -918,6 +934,63 @@ gdb_cmd_exception_pass(struct state *ks)
 	return -1;
 }
 
+static void
+gdb_cmd_continue(struct state *ks)
+{
+	if (strcmp("Stopped", getstatus(ks->threadid)) != 0) {
+		syslog(0, "gdbserver", "Process not stopped - can't activate breakpoints");
+		// Not really sure what to reply with here
+		return;
+	}
+
+	syslog(0, "gdbserver", "activating breakpoints");
+	dbg_activate_sw_breakpoints(ks);
+
+	// Block for the process to stop or receive a note
+	sendctl(ks->threadid, "startstop");
+
+	// Remove the breakpoint note so the process
+	// doesn't suicide
+	remove_note();
+
+	// Send code indicating we've hit a breakpoint
+	strcpy((char *)remcom_out_buffer, "S05");
+}
+
+static void
+gdb_cmd_single_step(struct state *ks)
+{
+	if (machdata == nil) {
+		syslog(0, "gdbserver", "machdata not set");
+		return;
+	}
+
+	if (machdata->foll == nil) {
+		syslog(0, "gdbserver", "machdata->foll not set");
+		return;
+	}
+
+	uint64_t foll[] = {0, 0, 0};
+	uint64_t pc = arch_get_pc(ks);
+	int nfoll = machdata->foll(cormap, pc, get_reg, foll);
+	if (nfoll < 0)
+		syslog(0, "gdbserver", "machdata->foll failed: %r");
+	else
+		syslog(0, "gdbserver", "machdata->foll succeeded, nfoll: %d", nfoll);
+
+	for (int i = 0; i < nfoll; i++) {
+		syslog(0, "gdbserver", "Setting single-step breakpoint at: %p", foll[i]);
+		dbg_set_sw_break(ks, foll[i]);
+	}
+
+	gdb_cmd_continue(ks);
+
+	for (int i = 0; i < nfoll; i++) {
```

Comment:
Do you mean re-add the breakpoint be just hit?  (The one we explicitly set, not the single step bps)
If so, what happens is that once you continue, gdb single steps, then sends a packet to re-add the breakpoint, before sending another 'c'ontinue packet.  So we don't need to do it ourselved.